show 'no connection' message in files tab OP#4068

individual-it opened the pull request:
In case there is no connection to OpenProject or workpackages could not be fetched, the user should see that information in the files tab and be able to navigate to the connection settings from there.

Currently this will always show "Error" as fetching workpackages is not implemented yet

individual-it commented on src/views/ProjectsTab.vue:
```diff
@@ -49,13 +53,33 @@ export default {
+			const req = {}
+			const url = generateUrl('/apps/integration_openproject/workpackages/' + fileId)
+			await axios.get(url, req).then((response) => {
+				if (!Array.isArray(response.data)) {
```
more validation need to be added later

individual-it commented on src/views/ProjectsTab.vue:
```diff
@@ -49,13 +53,33 @@ export default {
+
+		async fetchWorkpackages(fileId) {
+			const req = {}
+			const url = generateUrl('/apps/integration_openproject/workpackages/' + fileId)
```
this route is not implemented yet, so it will always return `200` with the HTML content of the dashboard

phil-davis commented on tests/jest/views/ProjectsTab.spec.js:
```diff
@@ -0,0 +1,71 @@
+			await localVue.nextTick()
+			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeTruthy()
+		})
+		it('does not shows the empty content message during "loading" state', async () => {
```
```suggestion
		it('does not show the empty content message during "loading" state', async () => {
```

individual-it replied:
fixed

phil-davis reviewed (approved):
LGTM - I only found one tiny grammar error in a test description!

individual-it commented:
OP#4068

kiranparajuli589 commented on src/components/tab/EmptyContent.vue:
```diff
@@ -4,18 +4,49 @@
+			{{ emptyContentMessage }}
+		</div>
+		<br>
+		<div v-if="state === 'no-token' || state === 'error'" class="connect-button">
```
```suggestion
		<div v-if="['error', 'no-token].includes(state)" class="connect-button">
```
you can also use this logic from the `computed` section, just to make HTML clean
```
<div v-if="someComputedMethod" class="connect-button">
```

individual-it replied:
done

kiranparajuli589 commented on src/components/tab/EmptyContent.vue:
```diff
@@ -4,18 +4,49 @@
+			} else if (this.state === 'error') {
+				return t('integration_openproject', 'Error connecting to OpenProject')
+			} else if (this.state === 'ok') {
+				return t('integration_openproject', 'No OpenProject links yet')
```
```suggestion
				return t('integration_openproject', 'No workspaces linked yet')
```

individual-it replied:
done

kiranparajuli589 commented on src/views/ProjectsTab.vue:
```diff
@@ -49,13 +53,33 @@ export default {
+		async fetchWorkpackages(fileId) {
+			const req = {}
+			const url = generateUrl('/apps/integration_openproject/workpackages/' + fileId)
+			await axios.get(url, req).then((response) => {
```
if you're using `then` & `catch`, await may not be necessary

eneiluj replied:
I agree with @kiranparajuli589 . It's clearer to choose between `await` in a try/catch block and `.then().catch()` without `await`.

As async/await is used all the way up to `projectTab.js::mount()`, this could just become
``` js
try {
    const response = await axios.get(url, req)
    if (!Array.isArray(response.data)) {
        ...
    }
} catch (error) {
    ...
}
```

individual-it replied:
done

kiranparajuli589 commented on src/views/ProjectsTab.vue:
```diff
@@ -49,13 +53,33 @@ export default {
 		 * Reset the current view to its default state
 		 */
 		resetState() {
 			this.loading = false
```
this can also be removed

individual-it replied:
I think we need that line because we need to fetch the information every time an other file is selected

kiranparajuli589 replied:
since `loading` from `data` section is removed, now `this.loading` means nothing. it will only warn you about the variable not found.

individual-it replied:
:facepalm: I thought you commented on `await this.fetchWorkpackages(this.fileInfo.id)`
you are totally right, I've removed the line

kiranparajuli589 commented on tests/jest/views/ProjectsTab.spec.js:
```diff
@@ -0,0 +1,71 @@
+		})
+	})
+	describe('empty message', () => {
+		it.each(['no-token', 'ok', 'error'])('shows the empty message on states != loading', async (state) => {
```
```suggestion
		it.each(['no-token', 'ok', 'error'])('should show the empty message when state is other than loading', async (state) => {
```

individual-it replied:
done

eneiluj reviewed (commented):
LGTM.

I agree with all comments from @kiranparajuli589.

individual-it commented:
@kiranparajuli589 @eneiluj thank you for the reviews, I have addressed the comments. 
Please review again
For the easy to re-review I've created a new commit - when this is ready to be merged I will squash it.

kiranparajuli589 reviewed (approved):
lgtm :+1:
